testing/test_step_saving.py

```diff
@@ -0,0 +1,232 @@
+    assert not os.path.exists(create_some_step3_path(tmpdir))
+
+
+def test_resumable_pipeline_transform_should_not_save_steps(tmpdir: LocalPath):
+    p = ResumablePipeline([
+        (SOME_STEP_1, MultiplyBy(multiply_by=2)),
+        (PIPELINE_2, ResumablePipeline([
+            (SOME_STEP_2, MultiplyBy(multiply_by=4)),
+            (CHECKPOINT, PickleCheckpointStep(cache_folder=tmpdir)),
+            (SOME_STEP_3, MultiplyBy(multiply_by=6)),
+        ]))
+    ], cache_folder=tmpdir)
+    p.name = ROOT
+
+    outputs = p.transform(
+        np.array(range(10))
+    )
+
+    assert np.array_equal(outputs, EXPECTED_OUTPUTS)
+    assert not os.path.exists(create_root_path(tmpdir))
+    assert not os.path.exists(create_pipeline2_path(tmpdir))
+    assert not os.path.exists(create_some_step1_path(tmpdir))
+    assert not os.path.exists(create_some_step2_path(tmpdir))
+    assert not os.path.exists(create_some_step3_path(tmpdir))
+    assert not os.path.exists(create_some_checkpoint_path(tmpdir))
+
+
+def test_resumable_pipeline_fit_should_save_all_fitted_pipeline_steps(tmpdir: LocalPath):
+    p = ResumablePipeline([
+        (SOME_STEP_1, MultiplyBy(multiply_by=2)),
+        (PIPELINE_2, ResumablePipeline([
+            (SOME_STEP_2, MultiplyBy(multiply_by=4)),
+            (CHECKPOINT, PickleCheckpointStep(cache_folder=tmpdir)),
+            (SOME_STEP_3, MultiplyBy(multiply_by=6)),
+        ]))
+    ], cache_folder=tmpdir)
+    p.name = ROOT
+
+    p = p.fit(
+        np.array(range(10)),
+        np.array(range(10))
+    )
+
+    assert os.path.exists(create_root_path(tmpdir))
+    assert os.path.exists(create_some_step1_path(tmpdir))
+    assert os.path.exists(create_pipeline2_path(tmpdir))
+    assert os.path.exists(create_some_step2_path(tmpdir))
+    assert os.path.exists(create_some_checkpoint_path(tmpdir))
+    assert not os.path.exists(create_some_step3_path(tmpdir))
+
+
+def test_resumable_pipeline_fit_transform_should_load_all_pipeline_steps(tmpdir: LocalPath):
+    # Given
+    p = given_saved_pipeline(tmpdir)
+
+    # When
+    p, outputs = p.fit_transform(
+        np.array(range(10)),
+        np.array(range(10))
+    )
+
+    # Then
+    assert np.array_equal(outputs, EXPECTED_OUTPUTS)
+
+
+def test_resumable_pipeline_transform_should_load_all_pipeline_steps(tmpdir: LocalPath):
+    # Given
+    p = given_saved_pipeline(tmpdir)
+
+    # When
+    outputs = p.transform(
+        np.array(range(10))
+    )
+
+    # Then
+    assert np.array_equal(outputs, EXPECTED_OUTPUTS)
+
+
+def test_resumable_pipeline_fit_should_load_all_pipeline_steps(tmpdir: LocalPath):
+    # Given
+    p = given_saved_pipeline(tmpdir)
+
+    # When
+    p = p.fit(
+        np.array(range(10)),
+        np.array(range(10))
+    )
+
+    # Then
+    assert p[SOME_STEP_1].multiply_by == 2
+    assert p[PIPELINE_2][SOME_STEP_2].multiply_by == 4
+    assert p[PIPELINE_2][SOME_STEP_3].multiply_by == 6
+
+
+def given_saved_pipeline(tmpdir):
```

Comment:
Badly named test, the test is so long that it's hard to see what it does. It seems that there could be a for loop in the test or parametrizxation perhaps, I didn't read it.

Reply:
I looped on each sub step